File: src/test/rs/emulate/legacy/archive/ArchiveTest.java
package rs.emulate.legacy.archive;

import static org.junit.Assert.assertEquals;

import java.io.IOException;
import java.util.Arrays;

import org.junit.Test;

import rs.emulate.shared.util.DataBuffer;

/**
 * Tests that the {@link ArchiveCodec#encode} and {@link ArchiveCodec#decode} methods are functioning correctly.
 * 
 * @author Major
 */
public final class ArchiveTest {

	/**
	 * The buffer used for testing, containing 15 bytes.
	 */
	private static final DataBuffer TEST_BUFFER = DataBuffer.wrap(
			new byte[] { 32, 78, 107, -30, 29, -81, -49, 113, 117, 51, 26, -30, -96, -34, 68 }).asReadOnlyBuffer();

	/**
	 * The archive entry identifier used in the test.
	 */
	private static final int TEST_IDENTIFIER = ArchiveUtils.hash("test");

	/**
	 * The ArchiveEntry to test.
	 */
	private static final ArchiveEntry ENTRY = new ArchiveEntry(TEST_IDENTIFIER, TEST_BUFFER);

	/**
	 * Executes the test.
	 * 
	 * @throws IOException If there is an error encoding or decoding the archive.
	 */
	@Test
	public void test() throws IOException {
		Archive archive = new Archive(Arrays.asList(ENTRY));

		DataBuffer encoded = ArchiveCodec.encode(archive, CompressionType.ARCHIVE_COMPRESSION);
		Archive decoded = ArchiveCodec.decode(encoded);

		assertEquals(archive, decoded);

		encoded = ArchiveCodec.encode(archive, CompressionType.ENTRY_COMPRESSION);
		decoded = ArchiveCodec.decode(encoded);

		assertEquals(archive, decoded);
	}

}

File: legacy/src/test/kotlin/rs/emulate/legacy/archive/ArchiveTest.kt
package rs.emulate.legacy.archive

import io.netty.buffer.Unpooled
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class ArchiveTest {

    @Test
    fun `end-to-end archive serialization using archive compression`() {
        val archive = Archive(listOf(ArchiveEntry(TEST_IDENTIFIER, TEST_BUFFER)))

        val encoded = ArchiveCodec.encode(archive, CompressionType.ARCHIVE_COMPRESSION)
        val decoded = ArchiveCodec.decode(encoded)

        assertEquals(archive, decoded)
    }

    @Test
    fun `end-to-end archive serialization using entry compression`() {
        val archive = Archive(listOf(ArchiveEntry(TEST_IDENTIFIER, TEST_BUFFER)))

        val encoded = ArchiveCodec.encode(archive, CompressionType.ENTRY_COMPRESSION)
        val decoded = ArchiveCodec.decode(encoded)

        assertEquals(archive, decoded)
    }

    @Test
    fun `archive entry name hashing`() {
        assertEquals(11943852, "test".entryHash())
    }

    companion object {

        private val TEST_BUFFER = Unpooled.wrappedBuffer(
            byteArrayOf(32, 78, 107, -30, 29, -81, -49, 113, 117, 51, 26, -30, -96, -34, 68)
        )

        private val TEST_IDENTIFIER = "test".entryHash()

    }

}
